installinator/src/write.rs

```diff
@@ -0,0 +1,173 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::time::Duration;
+
+use anyhow::{Context, Result};
+use buf_list::BufList;
+use bytes::Buf;
+use camino::{Utf8Path, Utf8PathBuf};
+use installinator_common::{CompletionEventKind, ProgressEventKind};
+use omicron_common::update::ArtifactHashId;
+use tokio::{io::AsyncWriteExt, sync::mpsc, time::Instant};
+
+use crate::reporter::ReportEvent;
+
+#[derive(Clone, Debug)]
+pub(crate) struct WriteDestination {
+    pub(crate) host_phase_2: Utf8PathBuf,
+    pub(crate) control_plane: Utf8PathBuf,
+}
+
+impl WriteDestination {
+    pub(crate) fn in_directory(dir: &Utf8Path) -> Result<Self> {
+        std::fs::create_dir_all(&dir)
+            .with_context(|| format!("error creating directories at {dir}"))?;
+        Ok(Self {
+            host_phase_2: dir.join("host_phase_2.bin"),
+            control_plane: dir.join("control_plane.bin"),
+        })
+    }
+}
+
+pub(crate) async fn write_artifact(
+    log: &slog::Logger,
+    artifact_id: &ArtifactHashId,
+    artifact: BufList,
+    destination: &Utf8Path,
+    event_sender: &mpsc::Sender<ReportEvent>,
+) {
+    let mut attempt = 0;
+
+    loop {
+        attempt += 1;
+        slog::info!(
+            log,
+            "writing artifact ({} bytes) to {destination} (attempt {attempt})",
+            artifact.num_bytes();
+            "artifact_id" => ?artifact_id
+        );
+        match write_artifact_impl(
+            attempt,
+            artifact_id,
+            artifact.clone(),
+            destination,
+            event_sender,
+        )
+        .await
+        {
+            Ok(()) => break,
+            Err(error) => {
+                slog::info!(log, "{error:?}"; "artifact_id" => ?artifact_id);
+                // Give it a short break, then keep trying.
+                tokio::time::sleep(Duration::from_secs(5)).await;
```

Comment:
I think this is more of a note for @andrewjstone and me: If installinator is willing to retry forever (which might be reasonable? or maybe we should give up after N attempts?), we should expose a way to cancel a running update from wicket; currently there's no way to cancel or restart an update that's running until it finishes.

Reply:
> In this case I think I'd suggest a reboot of the host, although I realize that would trigger a pull of at the phase 2 trampoline again, which is quite slow.

This is the (admittedly blunt) path I was thinking. "Cancel the update and start over."